pymongo/mongo_client.py

```diff
@@ -2304,6 +2205,159 @@ def __exit__(
+        )
+
+        self._func = func
+        self._bulk = bulk
+        self._multiple_retries = _csot.get_timeout() is not None
+        self._session = session
+        self._is_write = is_write
+        self._retryable = retryable and self._in_transaction()
+        self._read_pref = read_pref
+        self._server_selector = writable_server_selector if is_write else read_pref
+        self._address = address
+        self._server: Server = None
+
+    def run(self) -> T:
+        """Runs the supplied func() and attempts a retry
+
+        :raises self._last_error: Exception raised
+        :return: Result of the func() call
+        :rtype: T
+        """
+        if self._in_transaction() and self._is_write:
+            self._session._start_retryable_write()
+            if self._bulk:
+                self._bulk.started_retryable_write = True
+
+        while True:
+            self._check_last_error(check_csot=True)
+            try:
+                self._server = self._client._select_server(
+                    self._server_selector, self._session, address=self._address
+                )
+                return self._write() if self._is_write else self._read()
+            except ServerSelectionTimeoutError:
+                # The application may think the write was never attempted
+                # if we raise ServerSelectionTimeoutError on the retry
+                # attempt. Raise the original exception instead.
+                self._check_last_error()
+                # A ServerSelectionTimeoutError error indicates that there may
+                # be a persistent outage. Attempting to retry in this case will
+                # most likely be a waste of time.
+                raise
+            except ConnectionFailure as exc:
+                if self._is_not_retry_eligible():
+                    raise
+                self._retrying = True
+                self._last_error = exc
+            except OperationFailure as exc:
+                if self._is_not_retry_eligible() or exc.code not in helpers._RETRYABLE_ERROR_CODES:
+                    raise
+                self._retrying = True
+                self._last_error = exc
+            except PyMongoError as exc:
+                # Only execute additional handling when on a write call with retries
+                if not self._retryable or not self._is_write:
+                    raise
+                assert self._session
+                if exc.has_error_label("RetryableWriteError"):
+                    self._session._unpin()
+                elif self._is_retrying() and not self._multiple_retries:
+                    if exc.has_error_label("NoWritesPerformed") and self._last_error:
+                        raise self._last_error from exc
+                    else:
+                        raise
+                if self._bulk:
+                    self._bulk.retrying = True
```

Comment:
self._bulk.retrying needs to be set when retrying any exception not just PyMongoError.